source/common/upstream/ring_hash_lb.cc

```diff
@@ -138,13 +174,22 @@ void RingHashLoadBalancer::Ring::create(
 #endif
 }
 
-void RingHashLoadBalancer::refresh(uint32_t priority) {
-  while (per_priority_state_.size() < priority + 1) {
-    per_priority_state_.push_back(PerPriorityStatePtr{new PerPriorityState});
+void RingHashLoadBalancer::refresh() {
+  // Note that we only compute global panic on host set refresh. Given that the runtime setting will
+  // rarely changer, this is a reasonable compromise to avoid creating multiple rings when we only
+  // need to create one for LB.
+  RingConstSharedPtr new_ring;
+  const auto& host_set = chooseHostSet();
+  if (isGlobalPanic(host_set, runtime_)) {
+    new_ring = std::make_shared<Ring>(config_, host_set.hosts());
+    factory_->global_panic_ = true;
```

Comment:
Is there a possible race with this and the `current_ring_` update below?

Reply:
I used atomic so it should be OK, but yes it's possible it would be out of sync. I will just change it to be covered by the same lock.